sw/device/lib/base/memory_hardened.c

```diff
@@ -0,0 +1,670 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/memory_hardened.h"
+
+#include "sw/device/lib/base/hardened.h"
+
+static inline hardened_bool_t true_if_zero(uintptr_t val) {
+  /**
+   * We need a form where skipping instruction will not allow to
+   * produce kHardenedBoolTrue, on RV32 try to get 'snez' instead of 'seqz'
+   * as snez will transform non-zero reg into non-zero, so skipping it
+   * won't cause an inversion of value.
+   * Constant 7 is choosen to prevent input values in 0xfffffff0 - 0xffffffff
+   * range to be converted proper constant by skipping snez & neg.
+   */
+  uintptr_t mask = -(val * 7 != 0); /* val=0 -> 0, val!=0 => 0xfffffff */
+  return (hardened_bool_t)((mask & (uintptr_t)kHardenedBoolFalse) |
+                           ((~mask) & (uintptr_t)kHardenedBoolTrue));
+}
+
+/**
+ * Define machine word and alignment
+ */
+typedef uintptr_t word_t;
+#define WORD_SIZE sizeof(word_t)
+#define WORD_MASK (WORD_SIZE - 1)
+
+static inline size_t min(size_t a, size_t b) { return (a < b) ? a : b; }
+
+/**
+ * The 32 bit LFSR whose maximum length feedback polynomial is represented
+ * as X^32 + X^22 + X^2 + X^1 + 1 will produce 2^32-1 PN sequence.
+ * This LFSR can be initialized with 0,  but can't be initialized with
+ * 0xFFFFFFFF
+ */
+static uint32_t lfsr32(uint32_t seed) {
+  uint32_t bit = (seed ^ (seed >> 1) ^ (seed >> 21) ^ (seed >> 31)) & 1;
+  return (seed << 1) | (bit ^ 1);
+}
+
+/**
+ * Swap byte order between big-endian and little-endian
+ * @param x	Input word
+ * @return x in different endianess
+ */
+static inline uint32_t bswap32(uint32_t x) { return __builtin_bswap32(x); }
+
+/**
+ * Division by multiplication on mod 2^32 inverse.
+ * Result can be larger than interfer division
+ *
+ * @returns ~( p / (2^32/inv))
+ */
+static inline uint32_t div_by_mul(uint32_t p, uint32_t inv) {
+  return (inv) ? (uint32_t)(((uint64_t)p * (uint64_t)inv) >> 32) : p;
```

Comment:
this code specifically avoids 64-bit operations - it compiles to just ```mulhi ``` and that was the reason for this, as otherwise division will be faster.

Reply:
yes, I also have very efficient implementation of 64 bit by 32 bit division which is not in libgcc but very handy for big-number division. it was based on idea in libgcc, but resulted in faster code.